tsl/src/gapfill/exec.c

```diff
@@ -178,6 +189,182 @@ gapfill_state_create(CustomScan *cscan)
+	ListCell   *lc;
+
+	const char *op1,
+			   *op2;
+
+	if (GAPFILL_START == boundary)
+	{
+		op1 = ">";
+		op2 = ">=";
+	}
+	else
+	{
+		op1 = "<";
+		op2 = "<=";
+	}
+
+	/*
+	 * if the second argument to time_bucket_gapfill is not a column reference
+	 * we cannot match WHERE clause to the time column, we don't check for Var
+	 * directly though cause it might be wrapped in a cast
+	 */
+	vars = pull_var_clause(lsecond(func->args), 0);
+	if (list_length(vars) != 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid ts: need to refer to column if no start or finish is supplied")));
+
+	foreach(lc, (List *) jt->quals)
+	{
+		OpExpr	   *opexpr = lfirst(lc);
+
+		/*
+		 * check if left side of OpExpr is Var and matches Var passed in as
+		 * 2nd arg to time_bucket_gapfill call
+		 */
+		if (IsA(lfirst(lc), OpExpr) &&IsA(linitial(opexpr->args), Var) &&var_equal(linitial(vars), linitial(opexpr->args)))
+		{
+			char	   *oprname = get_oprname(opexpr->opno);
+
+			if (strcmp(oprname, op1) == 0 || strcmp(oprname, op2) == 0)
+			{
+				Datum		arg_value;
+				bool		isnull;
+				Expr	   *expr = lsecond(opexpr->args);
+
+				/*
+				 * add an explicit cast here to the type we require if
+				 * necessary
+				 */
+				if (exprType(lsecond(opexpr->args)) != state->gapfill_typid)
+				{
+					Oid			cast_oid = get_castfunc(exprType(lsecond(opexpr->args)), state->gapfill_typid);
+
+					expr = (Expr *) makeFuncExpr(cast_oid, state->gapfill_typid, list_make1(expr), InvalidOid, InvalidOid, 0);
+				}
+
+				arg_value = gapfill_exec_expr(state, expr, &isnull);
+
+				/*
+				 * if the expression evaluates to NULL we keep looking
+				 */
+				if (isnull)
+					continue;
+
+				/*
+				 * for gapfill start we pass the value through time_bucket to
+				 * align with chunk boundary
+				 */
+				if (GAPFILL_START == boundary)
+				{
+					int64		value = gapfill_datum_get_internal(arg_value, state->gapfill_typid);
+
+					/*
+					 * if we are looking for the start boundary we adjust the
+					 * value when greater than is used as expression because
+					 * the gapfill check is greater or equal so we add 1 here
+					 * for the greater case.
+					 */
+					if (strcmp(oprname, ">") == 0)
+						value += 1;
+
+					arg_value = DirectFunctionCall2(ts_int64_bucket, Int64GetDatum(state->gapfill_period), Int64GetDatum(value));
+				}
+
+				*value = gapfill_datum_get_internal(arg_value, state->gapfill_typid);
+
+				/*
+				 * if we are looking for the end boundary we adjust the value
+				 * when less or equal is used as expression because the
+				 * gapfill check is less than so we add 1 here for the equal
+				 * case.
+				 */
+				if (GAPFILL_END == boundary && strcmp(oprname, "<=") == 0)
+					*value += 1;
+
+				return;
+			}
+		}
+	}
+
+	ereport(ERROR,
+			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+			 errmsg("invalid %s: could not deduce boundary from WHERE clause", boundary == GAPFILL_START ? "start" : "finish")));
```

Comment:
This error message, and all the others also seems like it should refer  `gapfill` directly: right now the user sees something like `invalid start: could not deduce boundary from WHERE clause` with no indication _what_ boundary we're trying to deduce, or why. (this one in particular seems like it could also use an `errhint` indicating that they can add args to `gapfill` or to the WHERE clause)

Reply:
added an errorhint where appropriate and also added function name to error messages